src/TestSuite/IntegrationTestTrait.php

```diff
@@ -1201,7 +1227,39 @@ public function assertCookieEncrypted($expected, $name, $encrypt = 'aes', $key =
      */
     public function assertFileResponse($expected, $message = '')
     {
-        $this->assertThat(null, new FileSent($this->_response), $message);
-        $this->assertThat($expected, new FileSentAs($this->_response), $message);
+        $verboseMessage = $this->extractVerboseMessage($message);
+        $this->assertThat(null, new FileSent($this->_response), $verboseMessage);
+        $this->assertThat($expected, new FileSentAs($this->_response), $verboseMessage);
+    }
+
+    /**
+     * Inspect controller viewVars to extract possible causes of the failed assertion
+     *
+     * @param null $message
+     */
+    protected function extractVerboseMessage($message = null)
+    {
+        if ($this->_exception instanceof \Exception) {
+            $message .= $this->extractExceptionMessage($this->_exception);
+        }
+        if ($this->viewVariable('error') instanceof \Exception) {
```

Comment:
yes, https://github.com/cakephp/cakephp/blob/3.next/src/Error/ExceptionRenderer.php#L212 I actually did a test with a NotFoundException in my test project, I was thinking to also inspect the view variables, check if they are entities and do some more verbose message incase they have errors, many times I had a test not passing because a new validation error prevents the asserRedirect to happen...

Reply:
Ok. I think this is a good improvement without the entity munging, which could be done separately.